lib/Driver/ToolChains.cpp

```diff
@@ -1468,6 +1468,25 @@ void ToolChain::addLinkRuntimeLib(const ArgList &Args, ArgStringList &Arguments,
   Arguments.push_back(Args.MakeArgString(P));
 }
 
+void ToolChain::addInProcPluginServerPath(const StringRef PluginPath,
```

Comment:
I feel this is still incorrect. Say when building in macOS targeting Windows, `WindowsToolchain` is used, and it passes in the `bin` directory. But in macOS hosts it expect `lib/swift`.

Yes, as you said the entire `addPluginArguments` looks not doing the right thing.

Reply:
Yes. That's what I was asking in the comment above. It looks like more than this needs to move outside the `*ToolChains.cpp` and into `ToolChains.cpp` and make it compile conditionally depending on the platform, but I was not completely sure.

If I see things correctly I think this is what is happening:

- In `addPluginArguments`
  - Calculate a base path.
    - Darwin and Unix: Use `computeRuntimeResourcePathFromExecutablePath`
    - Windows: Use executable path (?)
   - Add the `-in-process-plugin-server-path` (this PR)
     - Darwin and Unix: Use a "host" directory. Each one with its platform extension.
     - Windows: Don't use extra directories. No "lib" prefix and ".dll" extension.
   - Add a `-plugin-path`
     - Darwin and Unix: Uses base path + "host/plugins"
     - Windows: Uses the executable path (?)
    - Add a local plugin path
      - Darwin and Unix: Add a "../../local/lib/swift/host/plugins" with an extra `-plugin-path`
      - Windows: nothing
- In `addExternalPluginFrontendArgs` and `addPlatformSpecificPluginFrontendArgs`
  - Only exist on Darwin
  - Does this need to be in fact for the Darwin compiler, and not for the Darwin targets?
 
I think all the code in `addPluginArguments` should be tied to the compiler host, and not the target, but I wanted to first have confirmation that this is in fact necessary. There might be something that I don't completely understand about the plugin compilation module. If you agree with the above and can answer about `addPlatformSpecificPluginFrontendArgs` I can try to do the changes. I knew this works in the case I was experiencing, probably because the `addPluginArguments` ended up being the same for Darwin and Unix after my changes, but I imagine Windows -> Unix will have its own share of problems as well.

Reply:
Thank you for doing this Daniel!

As for `addPluginArguments`, I believe everything you mentioned is correct. In Windows host, everything must be in the same directory as the `getDriver().getSwiftProgramPath()`.

As for `addExternalPluginFrontendArgs()`, this is for users of the macOS/iOS/etc SDKs, so it's basically tied to the target, and I believe it's fine this is in `DarwinToolchain`. That said, since Apple doesn't provide `swift-plugin-server` that runs other host platforms, maybe we should wrap this with `#if defined(__APPLE__)`.

Reply:
I hopefully matched the code as close as possible to the one in the `swift-driver`. Hopefully it is what you were asking for.